common/src/main/java/io/novafoundation/nova/common/utils/recyclerView/expandable/ExpandableItemAnimator.kt

```diff
@@ -0,0 +1,292 @@
+package io.novafoundation.nova.common.utils.recyclerView.expandable
+
+import android.animation.Animator
+import android.animation.AnimatorListenerAdapter
+import android.view.ViewPropertyAnimator
+import androidx.recyclerview.widget.RecyclerView.ViewHolder
+import androidx.recyclerview.widget.SimpleItemAnimator
+import io.novafoundation.nova.common.utils.recyclerView.expandable.animator.ExpandableAnimationItemState
+import io.novafoundation.nova.common.utils.recyclerView.expandable.animator.ExpandableAnimator
+
+/**
+ * Potential problems:
+ * - If in one time we will run add and move animation or remove and move animation - one of an animation will be cancelled
+ */
+abstract class ExpandableItemAnimator(
+    private val settings: ExpandableAnimationSettings,
+    private val expandableAnimator: ExpandableAnimator
+) : SimpleItemAnimator() {
+
+    private var preparedForAnimation = false
+
+    private val addAnimations = mutableMapOf<String, MutableList<ViewHolder>>() // Parent item to children
+    private val removeAnimations = mutableMapOf<String, MutableList<ViewHolder>>() // Parent item to children
+    private val moveAnimations = mutableListOf<ViewHolder>()
+
+    private val pendingAddAnimations = mutableSetOf<ViewHolder>()
+    private val pendingRemoveAnimations = mutableSetOf<ViewHolder>()
+    private val pendingMoveAnimations = mutableSetOf<ViewHolder>()
+
+    init {
+        addDuration = settings.duration
+        removeDuration = settings.duration
+        moveDuration = settings.duration
+
+        supportsChangeAnimations = false
+    }
+
+    /**
+     * Use this method before adapter.submitList() to prepare items for animation.
+     * Item animations will be skipped otherwise
+     */
+    fun prepareForAnimation() {
+        preparedForAnimation = true
+    }
+
+    override fun animateAdd(holder: ViewHolder): Boolean {
+        val notPreparedForAnimation = !preparedForAnimation
+        val notExpandableChildItem = holder !is ExpandableChildViewHolder || holder.expandableItem == null
+        if (notPreparedForAnimation || notExpandableChildItem) {
+            dispatchAddFinished(holder)
+            return true
+        }
+
+        val item = (holder as ExpandableChildViewHolder).expandableItem!!
+
+        // Reset move state helps clear translationY when animation is being to be canceled
+        if (pendingMoveAnimations.contains(holder)) {
+            holder.itemView.animate().cancel()
+
+            resetMoveState(holder)
+        }
+
+        if (pendingRemoveAnimations.contains(holder)) {
+            holder.itemView.animate().cancel()
+        } else {
+            preAddImpl(holder)
+        }
+
+        if (item.groupId !in addAnimations) addAnimations[item.groupId] = mutableListOf()
+        addAnimations[item.groupId]?.add(holder)
```

Comment:
We can use `MultiMapList` list here (might need to create `put` method similar to `MutableMultiMap<K, V>.put()`